erc1155/src/lib.rs

```diff
@@ -155,86 +148,115 @@ impl ERC1155TokenBase for ERC1155Token {
         }
 
-        for (i, ele) in ids.iter().enumerate() {
-            let amount = amounts[i];
-
-            let from_balance = self.balance_of(from, ele);
-
-            if from_balance < amount {
-                panic!("ERC1155: insufficient balance for transfer")
+        for (id, amount) in ids.iter().zip(amounts) {
+            if !self.can_transfer(from, id, *amount) {
+                panic!("ERC1155: all batch element should be transerfable");
             }
-
-            self.set_balance(from, ele, from_balance.saturating_sub(amount));
-            let to_balance = self.balance_of(to, ele);
-            self.set_balance(to, ele, to_balance.saturating_add(amount));
         }
+
+        ids.iter()
+            .enumerate()
+            .for_each(|(idx, x)| self.safe_transfer_from(from, to, x, amounts[idx]));
+    }
+
+    fn can_transfer(&self, from: &ActorId, id: &u128, amount: u128) -> bool {
+        from == &msg::source()
+            || from == &exec::origin()
+            || self.get_balance(&msg::source(), id) >= amount
     }
 }
 
 impl ExtendERC1155TokenBase for ERC1155Token {
-    fn owner_of(&self, id: &u128) -> bool {
+    fn is_owner_of(&self, id: &u128) -> bool {
         let owner = msg::source();
-
         self.balance_of(&owner, id) != 0
     }
 
-    fn owner_of_batch(&self, ids: &[u128]) -> bool {
-        for (_, ele) in ids.iter().enumerate() {
-            let res = self.owner_of(ele);
-            if !res {
+    fn is_owner_of_batch(&self, ids: &[u128]) -> bool {
+        for ele in ids {
+            if !self.is_owner_of(ele) {
                 return false;
             }
         }
-
         true
     }
 
-    fn mint(&mut self, account: &ActorId, id: &u128, amount: u128) {
+    fn mint(&mut self, account: &ActorId, id: &u128, amount: u128, meta: &Option<TokenMetadata>) {
         if account == &ZERO_ID {
             panic!("ERC1155: Mint to zero address")
         }
+        if let Some(metadata) = meta {
+            // Off you go...
+            if amount > 1 {
+                panic!("ERC1155: Mint metadata to a fungible token")
+            }
+            self.token_metadata.insert(*id, metadata.clone());
+        }
         let prev_balance = self.balance_of(account, id);
         self.set_balance(account, id, prev_balance.saturating_add(amount));
     }
 
-    fn mint_batch(&mut self, account: &ActorId, ids: &[u128], amounts: &[u128]) {
+    fn mint_batch(
+        &mut self,
+        account: &ActorId,
+        ids: &[u128],
+        amounts: &[u128],
+        meta: Vec<Option<TokenMetadata>>,
+    ) {
         if account == &ZERO_ID {
             panic!("ERC1155: Mint to zero address")
         }
 
         if ids.len() != amounts.len() {
             panic!("ERC1155: ids and amounts length mismatch")
         }
+        ids.iter()
+            .enumerate()
+            .for_each(|(idx, x)| self.mint(account, x, amounts[idx], &meta[idx]));
+    }
```

Comment:
The final step to avoid cloning.

```suggestion
        meta.into_iter()
            .enumerate()
            .for_each(|(i, meta)| self.mint(account, &ids[i], amounts[i], meta));
    }
```